File: mobile/test/place_detail_test.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:jiyidashi/api_client.dart';
import 'package:jiyidashi/place_detail_page.dart';
import 'package:jiyidashi/stage1_app.dart';

class _PlaceApi extends JiYiApiClient {
  _PlaceApi({
    this.detail,
    this.detailError,
    this.places = const [],
    this.pendingDetail,
    this.nextDetail,
  }) : super(baseUrl: 'http://place-detail.invalid/v1') {
    accessToken = 'token';
    authenticatedUserId = 'owner';
  }

  final Map<String, dynamic>? detail;
  final Object? detailError;
  final List<Map<String, dynamic>> places;
  final Completer<Map<String, dynamic>>? pendingDetail;
  final Map<String, dynamic>? nextDetail;
  int detailCalls = 0;

  @override
  Future<List<Map<String, dynamic>>> listPlaces({int limit = 100}) async => places;

  @override
  Future<Map<String, dynamic>> getPlaceDetail(
    String placeId, {
    int limit = 50,
    String? cursor,
  }) async {
    detailCalls += 1;
    if (pendingDetail != null) return pendingDetail!.future;
    if (detailError != null) throw detailError!;
    if (cursor != null) {
      return nextDetail ??
          {
            'place': detail!['place'],
            'visits': [_visit(id: 'visit-2', finalized: false)],
            'next_cursor': null,
          };
    }
    return detail!;
  }
}

Map<String, dynamic> _visit({
  String id = 'visit-1',
  bool finalized = true,
}) {
  return {
    'id': id,
    'arrived_at': '2026-09-20T08:00:00Z',
    'left_at': finalized ? '2026-09-20T09:00:00Z' : null,
    'duration_seconds': finalized ? 3600 : null,
    'confidence': 0.92,
    'source': 'GPS',
    'finalized_at': finalized ? '2026-09-20T10:00:00Z' : null,
    'visit_finalized': finalized,
  };
}

Map<String, dynamic> _place({
  List<Map<String, dynamic>> visits = const [],
  Object? cursor,
}) {
  return {
    'place': {
      'id': 'place-1',
      'name': '家',
      'name_source': 'USER',
      'address': '测试地址',
      'category': 'HOME',
      'visit_count': 2,
      'first_visited_at': '2026-09-18T08:00:00Z',
      'last_visited_at': '2026-09-20T08:00:00Z',
    },
    'visits': visits,
    'next_cursor': cursor,
  };
}

void main() {
  testWidgets('PlaceDetailPage exposes explicit loading state', (tester) async {
    final completer = Completer<Map<String, dynamic>>();
    final api = _PlaceApi(pendingDetail: completer);
    await tester.pumpWidget(
      MaterialApp(home: PlaceDetailPage(api: api, placeId: 'place-1')),
    );
    expect(find.text('正在读取地点详情…'), findsOneWidget);
  });

  testWidgets('PlaceDetailPage exposes error and retry state', (tester) async {
    final api = _PlaceApi(detailError: ApiException(404, 'PLACE_NOT_FOUND'));
    await tester.pumpWidget(
      MaterialApp(home: PlaceDetailPage(api: api, placeId: 'place-1')),
    );
    await tester.pumpAndSettle();
    expect(find.text('无法读取地点详情'), findsOneWidget);
    expect(find.text('PLACE_NOT_FOUND'), findsOneWidget);
    expect(find.widgetWithText(FilledButton, '重试'), findsOneWidget);
  });

  testWidgets('PlaceDetailPage exposes empty retained-visit state', (tester) async {
    final api = _PlaceApi(detail: _place());
    await tester.pumpWidget(
      MaterialApp(home: PlaceDetailPage(api: api, placeId: 'place-1')),
    );
    await tester.pumpAndSettle();
    expect(find.text('家'), findsOneWidget);
    expect(find.text('还没有到访记录'), findsOneWidget);
  });

  testWidgets('PlaceDetailPage distinguishes finalized and mutable visits and paginates', (tester) async {
    final api = _PlaceApi(
      detail: _place(
        visits: [_visit()],
        cursor: 'next-page',
      ),
    );
    await tester.pumpWidget(
      MaterialApp(home: PlaceDetailPage(api: api, placeId: 'place-1')),
    );
    await tester.pumpAndSettle();
    expect(find.text('已稳定的到访'), findsOneWidget);
    await tester.tap(find.widgetWithText(OutlinedButton, '加载更多'));
    await tester.pumpAndSettle();
    expect(find.text('仍在更新的到访'), findsOneWidget);
    expect(api.detailCalls, 2);
  });

  testWidgets(
    'PlaceDetailPage rejects non-bool finalized instead of rendering mutable fact',
    (tester) async {
      final malformed = _visit()..['visit_finalized'] = 'true';
      final api = _PlaceApi(detail: _place(visits: [malformed]));
      await tester.pumpWidget(
        MaterialApp(home: PlaceDetailPage(api: api, placeId: 'place-1')),
      );
      await tester.pumpAndSettle();

      expect(find.text('无法读取地点详情'), findsOneWidget);
      expect(find.text('仍在更新的到访'), findsNothing);
      expect(find.text('已稳定的到访'), findsNothing);
    },
  );

  testWidgets(
    'PlaceDetailPage rejects malformed next cursor atomically without appending visits',
    (tester) async {
      final initial = _place(visits: [_visit()], cursor: 'next-page');
      final malformedNext = _place(
        visits: [_visit(id: 'visit-2', finalized: false)],
        cursor: 123,
      );
      final api = _PlaceApi(detail: initial, nextDetail: malformedNext);

      await tester.pumpWidget(
        MaterialApp(home: PlaceDetailPage(api: api, placeId: 'place-1')),
      );
      await tester.pumpAndSettle();
      expect(find.text('已稳定的到访'), findsOneWidget);

      await tester.tap(find.widgetWithText(OutlinedButton, '加载更多'));
      await tester.pumpAndSettle();

      expect(find.text('部分到访记录加载失败'), findsOneWidget);
      expect(find.text('仍在更新的到访'), findsNothing);
      expect(find.textContaining('visit-2'), findsNothing);
    },
  );

  testWidgets('Timeline place entry opens PlaceDetailPage', (tester) async {
    final api = _PlaceApi(
      places: const [
        {'id': 'place-1', 'name': '家', 'address': '测试地址'}
      ],
      detail: _place(),
    );
    await tester.pumpWidget(MaterialApp(home: Scaffold(body: TimelinePage(api: api))));
    await tester.pumpAndSettle();
    expect(find.text('地点'), findsOneWidget);
    await tester.tap(find.text('家'));
    await tester.pumpAndSettle();
    expect(find.text('地点详情'), findsOneWidget);
    expect(find.text('还没有到访记录'), findsOneWidget);
  });
}

File: miniprogram/tests/placeDetail.test.ts
import assert from 'node:assert/strict'
import test from 'node:test'
import {
  assertPlaceDetailIdentity,
  buildPlaceDetailPath,
  buildPlacesPath,
  mergePlaceDetailPages,
  parsePlaceDetail,
  parsePlaceList,
  placeDetailRoute,
  placeListPresentation,
} from '../src/services/placeDetail'

function place(overrides: Record<string, unknown> = {}) {
  return {
    id: 'aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa',
    name: '家',
    automatic_name: '住宅',
    automatic_name_source: 'POI',
    user_name: '家',
    name_source: 'USER',
    name_revision: 3,
    name_updated_at: '2026-09-20T02:00:00Z',
    latitude: 39.9,
    longitude: 116.4,
    address: '测试地址',
    category: '住宅',
    first_visited_at: '2026-09-18T01:00:00Z',
    last_visited_at: '2026-09-20T01:00:00Z',
    visit_count: 2,
    is_user_named: true,
    ...overrides,
  }
}

function visit(overrides: Record<string, unknown> = {}) {
  return {
    id: 'bbbbbbbb-bbbb-4bbb-8bbb-bbbbbbbbbbbb',
    arrived_at: '2026-09-20T01:00:00Z',
    left_at: '2026-09-20T02:00:00Z',
    duration_seconds: 3600,
    confidence: 0.95,
    source: 'LOCATION_CLUSTER',
    finalized_at: '2026-09-20T02:05:00Z',
    visit_finalized: true,
    ...overrides,
  }
}

function page(overrides: Record<string, unknown> = {}) {
  return { place: place(), visits: [visit()], next_cursor: null, ...overrides }
}

test('place parser preserves authoritative USER display name over automatic candidate', () => {
  const parsed = parsePlaceList([place()])
  assert.equal(parsed.length, 1)
  assert.equal(parsed[0].name, '家')
  assert.equal(parsed[0].name_source, 'USER')
  assert.equal(parsed[0].user_name, '家')
  assert.equal(parsed[0].automatic_name, '住宅')
})

test('place detail rejects string finalized flag instead of rendering a mutable visit', () => {
  assert.throws(
    () => parsePlaceDetail(page({ visits: [visit({ visit_finalized: 'true' })] })),
    /visit_finalized/,
  )
})

test('place detail rejects malformed cursor before any pagination merge', () => {
  const current = parsePlaceDetail(page({ next_cursor: 'cursor-1' }))
  const snapshot = JSON.stringify(current)
  assert.throws(
    () => parsePlaceDetail(page({
      visits: [visit({ id: 'cccccccc-cccc-4ccc-8ccc-cccccccccccc', visit_finalized: false })],
      next_cursor: 123,
    })),
    /next_cursor/,
  )
  assert.equal(JSON.stringify(current), snapshot)
  assert.equal(current.visits.length, 1)
})

test('empty retained Visit page is a valid explicit empty state', () => {
  const parsed = parsePlaceDetail(page({ visits: [] }))
  assert.equal(parsed.visits.length, 0)
})

test('initial detail identity mismatch fails closed', () => {
  const parsed = parsePlaceDetail(page({ place: place({ id: 'dddddddd-dddd-4ddd-8ddd-dddddddddddd' }) }))
  assert.throws(
    () => assertPlaceDetailIdentity(parsed, 'aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa'),
    /地点身份/,
  )
})

test('pagination merge is atomic and rejects a different place identity', () => {
  const current = parsePlaceDetail(page({ next_cursor: 'cursor-1' }))
  const next = parsePlaceDetail(page({
    visits: [visit({ id: 'cccccccc-cccc-4ccc-8ccc-cccccccccccc', visit_finalized: false, finalized_at: null })],
    next_cursor: null,
  }))
  const merged = mergePlaceDetailPages(current, next)
  assert.equal(current.visits.length, 1)
  assert.equal(merged.visits.length, 2)
  assert.equal(merged.next_cursor, null)

  const wrongPlace = parsePlaceDetail(page({ place: place({ id: 'dddddddd-dddd-4ddd-8ddd-dddddddddddd' }) }))
  assert.throws(() => mergePlaceDetailPages(current, wrongPlace), /不同的地点/)
})

test('place detail path keeps owner implicit and forwards the opaque cursor', () => {
  const path = buildPlaceDetailPath(' aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa ', 20, 'opaque+/=')
  assert.match(path, /^\/location\/places\/aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa\?limit=20&cursor=/)
  assert.match(path, /opaque%2B%2F%3D/)
  assert.equal(path.includes('user_id='), false)
  assert.equal(buildPlacesPath(25), '/location/places?limit=25')
})

test('place detail navigation carries only opaque place id', () => {
  const route = placeDetailRoute('aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa')
  assert.equal(route, '/pages/place-detail/index?placeId=aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa')
  assert.equal(route.includes('user_id='), false)
})


test('successful empty Place list shows empty state without retry', () => {
  const state = placeListPresentation({
    loading: false,
    loaded: true,
    authenticated: true,
    placeCount: 0,
    error: '',
  })
  assert.deepEqual(state, {
    message: '还没有地点记录',
    showRetry: false,
  })
})

test('failed Place list request shows error with retry', () => {
  const state = placeListPresentation({
    loading: false,
    loaded: true,
    authenticated: true,
    placeCount: 0,
    error: '地点加载失败',
  })
  assert.deepEqual(state, {
    message: '地点加载失败',
    showRetry: true,
  })
})
